### PR TITLE
os: implement fuse interface for ObjectStore

### PR DESCRIPTION
**Pull request by liewegas**
- ceph-objectstore-tool --op fuse --mountpount mnt
- ceph-osd --osd-objectstore-fuse=true  (appears at $osd_data/fuse)

### CODE REVIEW

File: src/tools/ceph_objectstore_tool.cc [hunk 1]
```diff
@@ -2338,7 +2341,8 @@ int main(int argc, char **argv)
     usage(desc);
     myexit(1);
   }
-  if (op != "list" && op != "fsck" && vm.count("op") && vm.count("object")) {
+  if (op != "list" && op != "fsck" && op != "fuse" &&
```
Comment (dzafman) [inline src/tools/ceph_objectstore_tool.cc]: Checking for --op "fuse" (or "fsck" for that matter) are not needed here, unless they expect positional arguments.

**Comment by yuyuyu101** [comment]
Ah, what's the purpose to add a fuse volume as the backend of objectstore?

**Comment by liewegas** [comment]
With FileStore you can easily inspect the ondisk content and fix things if 
something goes wrong.  This lets you do the same thing with BlueStore or 
KStore or others .. with a standard and safe interface.

**Comment by liewegas** [comment]
passed a rados run

**Comment by liewegas** [comment]
https://github.com/ceph/ceph-qa-suite/pull/797

**Comment by liewegas** [comment]
Switched the test to be a workunit, run via teuthology instead of make check.

File: src/os/FuseStore.cc [hunk 2]
```diff
@@ -0,0 +1,1054 @@
...
+{
+  delete info;
+}
+
+/*
+ * / - root directory
+ * $cid/
+ * $cid/type - objectstore type
+ * $cid/bitwise_hash_start = lowest hash value
+ * $cid/bitwise_hash_end = lowest hash value
```
Comment (athanatos) [inline src/os/FuseStore.cc]: highest hash value?

File: src/os/FuseStore.cc [hunk 3]
```diff
@@ -0,0 +1,1054 @@
...
+      return FN_HASH_VAL;
+    goto do_dir;
+  }
+  return -ENOENT;
+
+ do_dir:
+  {
+    string o = v.front();
+    for (unsigned i = 0; i < o.size(); ++i)
+      if (o[i] == '|')
```
Comment (athanatos) [inline src/os/FuseStore.cc]: I'm not sure what's going on here.  How do we deal with objects with '/' in the name?  If this is how, how do we deal with objects with '|' in the name?

Comment (liewegas) [inline src/os/FuseStore.cc]: oops, that was an early hack.  the operator<< and parse now do 
proper escaping of / (and anything else weird), so this isn't needed.

File: src/os/FuseStore.cc [hunk 4]
```diff
@@ -0,0 +1,1054 @@
...
+  }
+
+  return 0;
+}
+
+static int os_readdir(const char *path,
+		      void *buf,
+		      fuse_fill_dir_t filler,
+		      off_t offset,
+		      struct fuse_file_info *fi)
```
Comment (athanatos) [inline src/os/FuseStore.cc]: We don't seem to be checking the return code from filler here or checking offset.  Does this work correctly for large directories?

File: src/os/FuseStore.cc [hunk 5]
```diff
@@ -0,0 +1,1054 @@
...
+
+  default:
+    return -EPERM;
+  }
+
+  if (!t.empty()) {
+    ceph::shared_ptr<ObjectStore::Sequencer> osr(
+      new ObjectStore::Sequencer("fuse"));
+    fs->store->apply_transaction(&*osr, t);
+    osr->flush();
```
Comment (athanatos) [inline src/os/FuseStore.cc]: This waits only for application, not commit.  If we are going to do that, shouldn't we share one sequencer for the whole fuse instance to ensure that updates are committed in order?

**Comment by liewegas** [comment]
okay,

- now waits for commit + apply
- shared sequencer
- readdir fixed to use fancier interface

**Comment by athanatos** [comment]
lgtm

**Pull request merged**